File: examples/companion-packages/dotnet-csharp-app/src/AgentXM.Examples.PawMatch.CSharp/PawMatchCli.cs
using System.Collections.Frozen;
using System.Collections.Immutable;
using System.CommandLine;
using System.CommandLine.Help;
using System.Diagnostics;
using AgentXM.Examples.TinyFlags.CSharp;
using TinyFlagsClient = AgentXM.Examples.TinyFlags.CSharp.TinyFlags;

namespace AgentXM.Examples.PawMatch.CSharp;

internal sealed class PawMatchCli
{
    private static readonly ImmutableArray<(string Flag, ImmutableArray<string> Tags)> AllFactors =
    [
        ("has-kids", ["good-with-kids", "gentle"]),
        ("quiet-home", ["mellow", "calm", "solo", "lap-cat"]),
        ("active", ["high-energy", "playful"]),
        ("first-time", ["gentle", "calm", "low-energy"]),
        ("multiple-pets", ["social"]),
        ("small-home", ["lap-cat", "solo", "low-energy"]),
    ];

    private static readonly FrozenSet<string> PopularityTags =
        new[] { "social", "good-with-kids", "calm", "mellow", "gentle" }
            .ToFrozenSet(StringComparer.Ordinal);

    private readonly TinyFlagsClient flags;
    private readonly EvaluationContext context;
    private readonly TextWriter @out;
    private readonly TextWriter err;

    public PawMatchCli(
        TinyFlagsClient? flags = null,
        EvaluationContext? context = null,
        TextWriter? @out = null,
        TextWriter? err = null)
    {
        this.flags = flags ?? PawMatchFlags.Create();
        this.context = context ?? new EvaluationContext(SessionId: Environment.UserName);
        this.@out = @out ?? Console.Out;
        this.err = err ?? Console.Error;
    }

    public RootCommand BuildRootCommand()
    {
        var speciesOption = new Option<string?>("--species")
        {
            Description = "Filter by species (dog|cat|rabbit|guinea-pig).",
        };
        var browse = new Command("browse", "Browse adoptable pets.");
        browse.Options.Add(speciesOption);
        browse.SetAction(result => Browse(result.GetValue(speciesOption)));

        var showPet = new Argument<string>("pet")
        {
            Description = "Pet slug (see 'pawmatch browse').",
        };
        var show = new Command("show", "Show details for a pet.");
        show.Arguments.Add(showPet);
        show.SetAction(result => Show(result.GetValue(showPet) ?? string.Empty));

        var hasKids = new Option<bool>("--has-kids") { Description = "Family with children." };
        var quietHome = new Option<bool>("--quiet-home") { Description = "Quiet, calm household." };
        var active = new Option<bool>("--active") { Description = "Active, outdoor lifestyle." };
        var firstTime = new Option<bool>("--first-time") { Description = "First-time pet adopter." };
        var multiplePets = new Option<bool>("--multiple-pets") { Description = "Other pets at home." };
        var smallHome = new Option<bool>("--small-home") { Description = "Small home or apartment." };
        var match = new Command("match", "Match pets to your lifestyle.");
        match.Options.Add(hasKids);
        match.Options.Add(quietHome);
        match.Options.Add(active);
        match.Options.Add(firstTime);
        match.Options.Add(multiplePets);
        match.Options.Add(smallHome);
        match.SetAction(result => Match(new MatchPreferences(
            HasKids: result.GetValue(hasKids),
            QuietHome: result.GetValue(quietHome),
            Active: result.GetValue(active),
            FirstTime: result.GetValue(firstTime),
            MultiplePets: result.GetValue(multiplePets),
            SmallHome: result.GetValue(smallHome))));

        var applyPet = new Argument<string>("pet") { Description = "Pet slug to apply for." };
        var apply = new Command("apply", "Start an adoption application.");
        apply.Arguments.Add(applyPet);
        apply.SetAction(result => Apply(result.GetValue(applyPet) ?? string.Empty));

        var fees = new Command("fees", "Show adoption fees.");
        fees.SetAction(_ => Fees());

        var returnSupport = new Command("return-support", "Return support information.");
        returnSupport.SetAction(_ => ReturnSupport());

        var charityArg = new Argument<string?>("charity")
        {
            Description = "Charity slug (optional — omit to list charities).",
            Arity = ArgumentArity.ZeroOrOne,
        };
        var focusOption = new Option<string?>("--focus")
        {
            Description = "Charity focus (all|shelters|rescue|policy).",
        };
        var openOption = new Option<bool>("--open")
        {
            Description = "Open the charity's donation URL in a browser.",
        };
        var donate = new Command("donate", "Browse animal-welfare charities to support.");
        donate.Arguments.Add(charityArg);
        donate.Options.Add(focusOption);
        donate.Options.Add(openOption);
        donate.SetAction(result => Donate(
            result.GetValue(charityArg),
            result.GetValue(focusOption),
            result.GetValue(openOption)));

        var root = new RootCommand("pawmatch — community pet adoption CLI.");
        root.Subcommands.Add(browse);
        root.Subcommands.Add(show);
        root.Subcommands.Add(match);
        root.Subcommands.Add(apply);
        root.Subcommands.Add(fees);
        root.Subcommands.Add(returnSupport);
        root.Subcommands.Add(donate);
        root.SetAction(result =>
        {
            new HelpAction().Invoke(result);
            return 0;
        });
        return root;
    }

    private int Browse(string? species)
    {
        var pets = Pets.FilterBySpecies(species).ToList();
        if (pets.Count == 0)
        {
            @out.WriteLine($"No adoptable pets found for species '{species}'.");
            return 0;
        }

        if (flags.Enabled(PawMatchFlags.LongStayHighlight, context))
        {
            var longStay = pets
                .Where(p => p.IsLongStay)
                .OrderByDescending(p => p.DaysInShelter)
                .FirstOrDefault();
            if (longStay is not null)
            {
                @out.WriteLine($"★ Featured long-stay friend — please consider {longStay.Name}!");
                @out.WriteLine();
            }
        }

        var style = Variants.Parse<PetCardStyle>(flags.Variant(PawMatchFlags.PetCardStyle, context));
        foreach (var pet in pets)
        {
            RenderPet(pet, style);
        }
        return 0;
    }

    private int Show(string slug)
    {
        var pet = Pets.FindBySlug(slug);
        if (pet is null)
        {
            err.WriteLine($"Unknown pet '{slug}'. Try 'pawmatch browse'.");
            return 1;
        }

        RenderPet(pet, PetCardStyle.Detailed);
        @out.WriteLine($"  Needs: {pet.Needs}");
        @out.WriteLine($"  Days in shelter: {pet.DaysInShelter}{(pet.IsLongStay ? " (long-stay)" : string.Empty)}");
        return 0;
    }

    private int Match(MatchPreferences preferences)
    {
        var strategy = Variants.Parse<MatchStrategy>(flags.Variant(PawMatchFlags.RecommendationStrategy, context));
        var depth = Variants.Parse<MatchDepth>(flags.Variant(PawMatchFlags.MatchQuizDepth, context));
        var factors = FactorsForDepth(depth);
        var userFlags = preferences.ToFlagSet();
        var wantsBuilder = new HashSet<string>(StringComparer.Ordinal);
        foreach (var factor in factors)
        {
            if (!userFlags.Contains(factor.Flag)) continue;
            foreach (var tag in factor.Tags)
            {
                wantsBuilder.Add(tag);
            }
        }
        var wants = wantsBuilder.ToFrozenSet(StringComparer.Ordinal);

        @out.WriteLine($"Strategy: {Variants.ToKebab(strategy)} • Quiz depth: {Variants.ToKebab(depth)} ({factors.Length} factor(s) considered)");
        if (preferences.IsEmpty)
        {
            @out.WriteLine("(no preference flags provided — try --has-kids --quiet-home --active --first-time)");
        }
        @out.WriteLine();

        IEnumerable<Pet> ranked = strategy switch
        {
            MatchStrategy.Popularity => Pets.All.OrderByDescending(p => p.Tags.Count(PopularityTags.Contains)),
            MatchStrategy.LongestStay => Pets.All.OrderByDescending(p => p.DaysInShelter),
            _ => Pets.All.OrderByDescending(p => p.Tags.Count(wants.Contains)),
        };

        foreach (var pet in ranked.Take(3))
        {
            @out.WriteLine($"  • {pet.Name} ({pet.Breed}, {pet.AgeYears}y) — {string.Join(", ", pet.Tags)}");
        }

        @out.WriteLine();
        @out.WriteLine("Adoption is a conversation — book a meet-and-greet to see if it's a fit.");
        return 0;
    }

    private int Apply(string slug)
    {
        var pet = Pets.FindBySlug(slug);
        if (pet is null)
        {
            err.WriteLine($"Unknown pet '{slug}'. Try 'pawmatch browse'.");
            return 1;
        }

        @out.WriteLine($"Adoption application for {pet.Name}");
        @out.WriteLine();
        @out.WriteLine("""
            Next steps:
              1. Application reviewed by an adoption counselor (1–2 days).
              2. Meet-and-greet scheduled at the shelter.
              3. 48-hour reflection period before finalizing.
              4. Take-home day — fees cover spay/neuter, vaccines, and microchip.
            """);

        if (flags.Enabled(PawMatchFlags.HomeCheckFollowup, context))
        {
            @out.WriteLine("  5. Two-week follow-up check from a counselor to see how you're settling in.");
        }

        @out.WriteLine();
        @out.WriteLine("Returns are always accepted, no questions asked.");

        if (flags.Enabled(PawMatchFlags.SuggestDonateAfterAdoption, context))
        {
            @out.WriteLine();
            @out.WriteLine($"If {pet.Name} brings you joy, please consider donating to a shelter:");
            @out.WriteLine("  pawmatch donate");
        }
        return 0;
    }

    private int Fees()
    {
        @out.WriteLine("Adoption fees");
        @out.WriteLine();
        if (flags.Enabled(PawMatchFlags.FeeBreakdownDetailed, context))
        {
            @out.WriteLine("""
                  Dog adoption — $150 total:
                    $60   spay / neuter surgery
                    $45   core vaccinations
                    $25   microchip and registration
                    $20   intake exam and deworming

                  Cat adoption — $90 total:
                    $50   spay / neuter surgery
                    $25   core vaccinations
                    $15   microchip and registration

                  Small animal — $35 total (intake exam + microchip).
                """);
        }
        else
        {
            @out.WriteLine("""
                  Dog adoption           $150
                  Cat adoption            $90
                  Small animal            $35

                  Fees cover spay/neuter, vaccines, and microchip.
                """);
        }

        @out.WriteLine();
        @out.WriteLine("No one is turned away for inability to pay — ask about our subsidy fund.");
        return 0;
    }

    private int ReturnSupport()
    {
        @out.WriteLine("""
            Return support

            If your adoption isn't working out, we're here to help.
              • Free behavior consultation with our trainers.
              • No-judgment returns at any time — your pet stays in our care.
              • Connections to low-cost vet and food assistance programs.

            Returning a pet is not a failure. Reach out as soon as you'd like support.
            """);
        return 0;
    }

    private int Donate(string? charitySlug, string? focusOverride, bool open)
    {
        var defaultFocus = Variants.Parse<DonateFocus>(flags.Variant(PawMatchFlags.DonateFocusDefault, context));
        var focus = focusOverride ?? Variants.ToKebab(defaultFocus);
        var showRatings = flags.Enabled(PawMatchFlags.ShowCharityRatings, context);

        if (charitySlug is not null)
        {
            var charity = Charities.FindBySlug(charitySlug);
            if (charity is null)
            {
                err.WriteLine($"Unknown charity '{charitySlug}'.");
                return 1;
            }

            if (open)
            {
                return OpenUrl(charity.Url);
            }

            RenderCharity(charity, showRatings);
            return 0;
        }

        var list = Charities.FilterByFocus(focus).ToList();
        @out.WriteLine($"Animal-welfare charities (focus: {focus})");
        @out.WriteLine();
        foreach (var charity in list)
        {
            RenderCharity(charity, showRatings);
            @out.WriteLine();
        }

        @out.WriteLine(Charities.Disclaimer);
        if (!showRatings)
        {
            @out.WriteLine("Ratings hidden — set show-charity-ratings to surface them inline.");
        }
        return 0;
    }

    private void RenderPet(Pet pet, PetCardStyle style)
    {
        var longStayBadge = pet.IsLongStay ? " ★" : string.Empty;
        switch (style)
        {
            case PetCardStyle.Compact:
                @out.WriteLine($"  {pet.Slug,-10} {pet.Name,-14} {pet.Species,-10} {pet.AgeYears}y{longStayBadge}");
                break;
            case PetCardStyle.Playful:
                @out.WriteLine($"  🐾 {pet.Name}{longStayBadge} — a {pet.AgeYears}-year-old {pet.Breed.ToLowerInvariant()} who is {string.Join(" & ", pet.Tags)}.");
                break;
            case PetCardStyle.Detailed:
            default:
                @out.WriteLine($"  {pet.Name}{longStayBadge}  [{pet.Slug}]");
                @out.WriteLine($"    {pet.Breed}, {pet.AgeYears} years old");
                @out.WriteLine($"    Tags: {string.Join(", ", pet.Tags)}");
                @out.WriteLine();
                break;
        }
    }

    private void RenderCharity(Charity charity, bool showRatings)
    {
        @out.WriteLine($"  {charity.Name}  [{charity.Slug}]");
        @out.WriteLine($"    Focus: {charity.Focus}");
        @out.WriteLine($"    {charity.Description}");
        @out.WriteLine($"    Donate: {charity.Url}");
        if (showRatings)
        {
            @out.WriteLine($"    Rating: {charity.RatingNote}");
        }
    }

    private int OpenUrl(string url)
    {
        try
        {
            ProcessStartInfo psi;
            if (OperatingSystem.IsWindows())
            {
                psi = new ProcessStartInfo(url) { UseShellExecute = true };
            }
            else if (OperatingSystem.IsMacOS())
            {
                psi = new ProcessStartInfo("open", url);
            }
            else if (OperatingSystem.IsLinux())
            {
                psi = new ProcessStartInfo("xdg-open", url);
            }
            else
            {
                psi = new ProcessStartInfo(url) { UseShellExecute = true };
            }
            Process.Start(psi);
            return 0;
        }
        catch (Exception ex) when (ex is System.ComponentModel.Win32Exception or InvalidOperationException or PlatformNotSupportedException)
        {
            err.WriteLine($"Unable to open browser ({ex.GetType().Name}). URL: {url}");
            return 1;
        }
    }

    private static ReadOnlySpan<(string Flag, ImmutableArray<string> Tags)> FactorsForDepth(MatchDepth depth)
    {
        var take = depth switch
        {
            MatchDepth.Short => 2,
            MatchDepth.Thorough => 6,
            _ => 4,
        };
        return AllFactors.AsSpan()[..take];
    }
}

File: examples/companion-packages/dotnet-fsharp-nuget-app/src/AgentXM.Examples.PawMatch.FSharp/PawMatchCli.fs
module AgentXM.Examples.PawMatch.FSharp.PawMatchCli

open System
open System.Diagnostics
open System.IO
open Argu
open AgentXM.Examples.TinyFlags.FSharp
open AgentXM.Examples.PawMatch.FSharp
open AgentXM.Examples.PawMatch.FSharp.Variants

let private allFactors: (string * string list) list =
    [ "has-kids", [ "good-with-kids"; "gentle" ]
      "quiet-home", [ "mellow"; "calm"; "solo"; "lap-cat" ]
      "active", [ "high-energy"; "playful" ]
      "first-time", [ "gentle"; "calm"; "low-energy" ]
      "multiple-pets", [ "social" ]
      "small-home", [ "lap-cat"; "solo"; "low-energy" ] ]

let private popularityTags =
    Set.ofList [ "social"; "good-with-kids"; "calm"; "mellow"; "gentle" ]

type BrowseArgs =
    | [<Unique>] Species of string

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Species _ -> "Filter by species (dog|cat|rabbit|guinea-pig)."

type ShowArgs =
    | [<MainCommand; ExactlyOnce; Last>] Pet of pet: string

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Pet _ -> "Pet slug (see 'pawmatch browse')."

type MatchArgs =
    | [<Unique>] Has_Kids
    | [<Unique>] Quiet_Home
    | [<Unique>] Active
    | [<Unique>] First_Time
    | [<Unique>] Multiple_Pets
    | [<Unique>] Small_Home

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Has_Kids -> "Family with children."
            | Quiet_Home -> "Quiet, calm household."
            | Active -> "Active, outdoor lifestyle."
            | First_Time -> "First-time pet adopter."
            | Multiple_Pets -> "Other pets at home."
            | Small_Home -> "Small home or apartment."

type ApplyArgs =
    | [<MainCommand; ExactlyOnce; Last>] Pet of pet: string

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Pet _ -> "Pet slug to apply for."

type DonateArgs =
    | [<MainCommand; Last>] Charity of charity: string
    | [<Unique>] Focus of string
    | [<Unique; CustomCommandLine("--open")>] Open_Browser

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Charity _ -> "Charity slug (optional — omit to list charities)."
            | Focus _ -> "Charity focus (all|shelters|rescue|policy)."
            | Open_Browser -> "Open the charity's donation URL in a browser."

type PawMatchCommand =
    | [<CliPrefix(CliPrefix.None)>] Browse of ParseResults<BrowseArgs>
    | [<CliPrefix(CliPrefix.None)>] Show of ParseResults<ShowArgs>
    | [<CliPrefix(CliPrefix.None)>] Match of ParseResults<MatchArgs>
    | [<CliPrefix(CliPrefix.None)>] Apply of ParseResults<ApplyArgs>
    | [<CliPrefix(CliPrefix.None)>] Fees
    | [<CliPrefix(CliPrefix.None); CustomCommandLine("return-support")>] ReturnSupport
    | [<CliPrefix(CliPrefix.None)>] Donate of ParseResults<DonateArgs>

    interface IArgParserTemplate with
        member this.Usage =
            match this with
            | Browse _ -> "Browse adoptable pets."
            | Show _ -> "Show details for a pet."
            | Match _ -> "Match pets to your lifestyle."
            | Apply _ -> "Start an adoption application."
            | Fees -> "Show adoption fees."
            | ReturnSupport -> "Return support information."
            | Donate _ -> "Browse animal-welfare charities to support."

type Cli =
    { Flags: TinyFlags
      Context: EvaluationContext
      Out: TextWriter
      Err: TextWriter }

let defaultContext () =
    { EvaluationContext.empty with
        SessionId = Some Environment.UserName }

let create () : Cli =
    { Flags = Flags.create ()
      Context = defaultContext ()
      Out = Console.Out
      Err = Console.Error }

let withWriters (out: TextWriter) (err: TextWriter) (cli: Cli) =
    { cli with Out = out; Err = err }

let private renderPet (cli: Cli) (pet: Pets.Pet) style =
    let longStayBadge = if Pets.isLongStay pet then " ★" else ""

    match style with
    | Compact ->
        cli.Out.WriteLine(
            sprintf "  %-10s %-14s %-10s %dy%s" pet.Slug pet.Name pet.Species pet.AgeYears longStayBadge
        )
    | Playful ->
        let tagText = String.concat " & " pet.Tags

        cli.Out.WriteLine(
            sprintf
                "  🐾 %s%s — a %d-year-old %s who is %s."
                pet.Name
                longStayBadge
                pet.AgeYears
                (pet.Breed.ToLowerInvariant())
                tagText
        )
    | Detailed ->
        cli.Out.WriteLine(sprintf "  %s%s  [%s]" pet.Name longStayBadge pet.Slug)
        cli.Out.WriteLine(sprintf "    %s, %d years old" pet.Breed pet.AgeYears)
        cli.Out.WriteLine(sprintf "    Tags: %s" (String.concat ", " pet.Tags))
        cli.Out.WriteLine()

let private renderCharity (cli: Cli) (charity: Charities.Charity) showRatings =
    cli.Out.WriteLine(sprintf "  %s  [%s]" charity.Name charity.Slug)
    cli.Out.WriteLine(sprintf "    Focus: %s" charity.Focus)
    cli.Out.WriteLine(sprintf "    %s" charity.Description)
    cli.Out.WriteLine(sprintf "    Donate: %s" charity.Url)
    if showRatings then
        cli.Out.WriteLine(sprintf "    Rating: %s" charity.RatingNote)

let private factorsForDepth depth =
    let take =
        match depth with
        | Short -> 2
        | Thorough -> 6
        | Standard -> 4

    allFactors |> List.truncate take

let browse (cli: Cli) (species: string option) =
    let pets = Pets.filterBySpecies species

    if List.isEmpty pets then
        let label = defaultArg species "<unspecified>"
        cli.Out.WriteLine(sprintf "No adoptable pets found for species '%s'." label)
        0
    else
        if cli.Flags |> TinyFlags.enabled Flags.LongStayHighlight cli.Context then
            let longStay =
                pets
                |> List.filter Pets.isLongStay
                |> List.sortByDescending (fun p -> p.DaysInShelter)
                |> List.tryHead

            match longStay with
            | Some pet ->
                cli.Out.WriteLine(sprintf "★ Featured long-stay friend — please consider %s!" pet.Name)
                cli.Out.WriteLine()
            | None -> ()

        let style =
            cli.Flags
            |> TinyFlags.variant Flags.PetCardStyle cli.Context
            |> parsePetCardStyle

        for pet in pets do
            renderPet cli pet style

        0

let show (cli: Cli) (slug: string) =
    match Pets.findBySlug slug with
    | None ->
        cli.Err.WriteLine(sprintf "Unknown pet '%s'. Try 'pawmatch browse'." slug)
        1
    | Some pet ->
        renderPet cli pet Detailed
        cli.Out.WriteLine(sprintf "  Needs: %s" pet.Needs)
        let longStaySuffix = if Pets.isLongStay pet then " (long-stay)" else ""
        cli.Out.WriteLine(sprintf "  Days in shelter: %d%s" pet.DaysInShelter longStaySuffix)
        0

let matchCommand (cli: Cli) (preferences: MatchPreferences.T) =
    let strategy =
        cli.Flags
        |> TinyFlags.variant Flags.RecommendationStrategy cli.Context
        |> parseMatchStrategy

    let depth =
        cli.Flags
        |> TinyFlags.variant Flags.MatchQuizDepth cli.Context
        |> parseMatchDepth

    let factors = factorsForDepth depth
    let userFlags = MatchPreferences.toFlagSet preferences

    let wants =
        factors
        |> List.fold
            (fun acc (factorFlag, tags) ->
                if Set.contains factorFlag userFlags then
                    tags |> List.fold (fun s t -> Set.add t s) acc
                else
                    acc)
            Set.empty

    cli.Out.WriteLine(
        sprintf
            "Strategy: %s • Quiz depth: %s (%d factor(s) considered)"
            (matchStrategyToKebab strategy)
            (matchDepthToKebab depth)
            (List.length factors)
    )

    if MatchPreferences.isEmpty preferences then
        cli.Out.WriteLine(
            "(no preference flags provided — try --has-kids --quiet-home --active --first-time)"
        )

    cli.Out.WriteLine()

    let scoreBy (tags: Set<string>) (pet: Pets.Pet) =
        pet.Tags |> List.filter tags.Contains |> List.length

    let ranked =
        match strategy with
        | Popularity -> Pets.all |> List.sortByDescending (scoreBy popularityTags)
        | LongestStay -> Pets.all |> List.sortByDescending (fun p -> p.DaysInShelter)
        | MatchQuiz -> Pets.all |> List.sortByDescending (scoreBy wants)

    for pet in ranked |> List.truncate 3 do
        cli.Out.WriteLine(
            sprintf "  • %s (%s, %dy) — %s" pet.Name pet.Breed pet.AgeYears (String.concat ", " pet.Tags)
        )

    cli.Out.WriteLine()
    cli.Out.WriteLine("Adoption is a conversation — book a meet-and-greet to see if it's a fit.")
    0

let apply (cli: Cli) (slug: string) =
    match Pets.findBySlug slug with
    | None ->
        cli.Err.WriteLine(sprintf "Unknown pet '%s'. Try 'pawmatch browse'." slug)
        1
    | Some pet ->
        cli.Out.WriteLine(sprintf "Adoption application for %s" pet.Name)
        cli.Out.WriteLine()
        cli.Out.WriteLine("Next steps:")
        cli.Out.WriteLine("  1. Application reviewed by an adoption counselor (1–2 days).")
        cli.Out.WriteLine("  2. Meet-and-greet scheduled at the shelter.")
        cli.Out.WriteLine("  3. 48-hour reflection period before finalizing.")
        cli.Out.WriteLine("  4. Take-home day — fees cover spay/neuter, vaccines, and microchip.")

        if cli.Flags |> TinyFlags.enabled Flags.HomeCheckFollowup cli.Context then
            cli.Out.WriteLine(
                "  5. Two-week follow-up check from a counselor to see how you're settling in."
            )

        cli.Out.WriteLine()
        cli.Out.WriteLine("Returns are always accepted, no questions asked.")

        if cli.Flags |> TinyFlags.enabled Flags.SuggestDonateAfterAdoption cli.Context then
            cli.Out.WriteLine()
            cli.Out.WriteLine(sprintf "If %s brings you joy, please consider donating to a shelter:" pet.Name)
            cli.Out.WriteLine("  pawmatch donate")

        0

let fees (cli: Cli) =
    cli.Out.WriteLine("Adoption fees")
    cli.Out.WriteLine()

    if cli.Flags |> TinyFlags.enabled Flags.FeeBreakdownDetailed cli.Context then
        cli.Out.WriteLine("  Dog adoption — $150 total:")
        cli.Out.WriteLine("    $60   spay / neuter surgery")
        cli.Out.WriteLine("    $45   core vaccinations")
        cli.Out.WriteLine("    $25   microchip and registration")
        cli.Out.WriteLine("    $20   intake exam and deworming")
        cli.Out.WriteLine()
        cli.Out.WriteLine("  Cat adoption — $90 total:")
        cli.Out.WriteLine("    $50   spay / neuter surgery")
        cli.Out.WriteLine("    $25   core vaccinations")
        cli.Out.WriteLine("    $15   microchip and registration")
        cli.Out.WriteLine()
        cli.Out.WriteLine("  Small animal — $35 total (intake exam + microchip).")
    else
        cli.Out.WriteLine("  Dog adoption           $150")
        cli.Out.WriteLine("  Cat adoption            $90")
        cli.Out.WriteLine("  Small animal            $35")
        cli.Out.WriteLine()
        cli.Out.WriteLine("  Fees cover spay/neuter, vaccines, and microchip.")

    cli.Out.WriteLine()
    cli.Out.WriteLine("No one is turned away for inability to pay — ask about our subsidy fund.")
    0

let returnSupport (cli: Cli) =
    cli.Out.WriteLine("Return support")
    cli.Out.WriteLine()
    cli.Out.WriteLine("If your adoption isn't working out, we're here to help.")
    cli.Out.WriteLine("  • Free behavior consultation with our trainers.")
    cli.Out.WriteLine("  • No-judgment returns at any time — your pet stays in our care.")
    cli.Out.WriteLine("  • Connections to low-cost vet and food assistance programs.")
    cli.Out.WriteLine()
    cli.Out.WriteLine("Returning a pet is not a failure. Reach out as soon as you'd like support.")
    0

let private openUrl (cli: Cli) (url: string) =
    try
        let psi =
            if OperatingSystem.IsWindows() then
                ProcessStartInfo(url, UseShellExecute = true)
            elif OperatingSystem.IsMacOS() then
                ProcessStartInfo("open", url)
            elif OperatingSystem.IsLinux() then
                ProcessStartInfo("xdg-open", url)
            else
                ProcessStartInfo(url, UseShellExecute = true)

        Process.Start(psi) |> ignore
        0
    with
    | :? System.ComponentModel.Win32Exception as ex ->
        cli.Err.WriteLine(sprintf "Unable to open browser (%s). URL: %s" (ex.GetType().Name) url)
        1
    | :? InvalidOperationException as ex ->
        cli.Err.WriteLine(sprintf "Unable to open browser (%s). URL: %s" (ex.GetType().Name) url)
        1
    | :? PlatformNotSupportedException as ex ->
        cli.Err.WriteLine(sprintf "Unable to open browser (%s). URL: %s" (ex.GetType().Name) url)
        1

let donate (cli: Cli) (charitySlug: string option) (focusOverride: string option) (openBrowser: bool) =
    let defaultFocus =
        cli.Flags
        |> TinyFlags.variant Flags.DonateFocusDefault cli.Context
        |> parseDonateFocus

    let focus = defaultArg focusOverride (donateFocusToKebab defaultFocus)
    let showRatings = cli.Flags |> TinyFlags.enabled Flags.ShowCharityRatings cli.Context

    match charitySlug with
    | Some slug ->
        match Charities.findBySlug slug with
        | None ->
            cli.Err.WriteLine(sprintf "Unknown charity '%s'." slug)
            1
        | Some charity ->
            if openBrowser then
                openUrl cli charity.Url
            else
                renderCharity cli charity showRatings
                0
    | None ->
        let list = Charities.filterByFocus focus
        cli.Out.WriteLine(sprintf "Animal-welfare charities (focus: %s)" focus)
        cli.Out.WriteLine()

        for charity in list do
            renderCharity cli charity showRatings
            cli.Out.WriteLine()

        cli.Out.WriteLine(Charities.disclaimer)

        if not showRatings then
            cli.Out.WriteLine("Ratings hidden — set show-charity-ratings to surface them inline.")

        0

let private preferencesOf (results: ParseResults<MatchArgs>) =
    { MatchPreferences.empty with
        HasKids = results.Contains Has_Kids
        QuietHome = results.Contains Quiet_Home
        Active = results.Contains Active
        FirstTime = results.Contains First_Time
        MultiplePets = results.Contains Multiple_Pets
        SmallHome = results.Contains Small_Home }

let runCommand (cli: Cli) (command: PawMatchCommand) =
    match command with
    | Browse args -> browse cli (args.TryGetResult BrowseArgs.Species)
    | Show args -> show cli (args.GetResult ShowArgs.Pet)
    | Match args -> matchCommand cli (preferencesOf args)
    | Apply args -> apply cli (args.GetResult ApplyArgs.Pet)
    | Fees -> fees cli
    | ReturnSupport -> returnSupport cli
    | Donate args ->
        donate
            cli
            (args.TryGetResult DonateArgs.Charity)
            (args.TryGetResult DonateArgs.Focus)
            (args.Contains DonateArgs.Open_Browser)

let createParser () =
    ArgumentParser.Create<PawMatchCommand>(programName = "pawmatch", errorHandler = ProcessExiter())

let run (cli: Cli) (argv: string[]) =
    let parser = createParser ()
    let results = parser.ParseCommandLine(argv, raiseOnUsage = false)

    if results.IsUsageRequested || results.GetAllResults() |> List.isEmpty then
        cli.Out.WriteLine(parser.PrintUsage(message = "pawmatch — community pet adoption CLI."))
        0
    else
        let commands = results.GetAllResults()

        match commands with
        | [ command ] -> runCommand cli command
        | _ ->
            cli.Err.WriteLine("Specify exactly one pawmatch subcommand.")
            1
